pkg/controller/azure/compute/aks.go

```diff
@@ -0,0 +1,493 @@
+	instance.Status.Endpoint = util.StringValue(cluster.Fqdn)
+	instance.Status.ProviderID = util.StringValue(cluster.ID)
+	instance.Status.State = util.StringValue(cluster.ProvisioningState)
+
+	if !instance.Status.IsReady() {
+		instance.Status.UnsetAllConditions()
+	}
+	instance.Status.SetReady()
+
+	return resultDone, r.Update(ctx, instance)
+}
+
+// _secret AKS connection credentials
+func (r *Reconciler) _secret(instance *azurecomputev1alpha1.AKSCluster, client azure.AKSClientsetAPI) (*corev1.Secret, error) {
+	creds, err := client.ListCredentials(instance.Spec.ResourceGroupName, instance.Status.ClusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := clientcmd.Load(*creds.Value)
+	if err != nil {
+		return nil, err
+	}
+
+	// cluster name - used as a context key
+	name := instance.Status.ClusterName
+
+	kubeContext, ok := config.Contexts[name]
+	if !ok {
+		return nil, fmt.Errorf("context configuration is not found for cluster: %s", name)
+	}
+	cluster, ok := config.Clusters[kubeContext.Cluster]
+	if !ok {
+		return nil, fmt.Errorf("cluster configuration is not found: %s", kubeContext.Cluster)
+	}
+	auth, ok := config.AuthInfos[kubeContext.AuthInfo]
+	if !ok {
+		return nil, fmt.Errorf("auth-info configuration is not found: %s", kubeContext.AuthInfo)
+	}
+
+	// create the connection secret with the credentials data now
+	connectionSecret := instance.ConnectionSecret()
+	data := make(map[string][]byte)
+	data[corev1alpha1.ResourceCredentialsSecretEndpointKey] = []byte(cluster.Server)
+	data[corev1alpha1.ResourceCredentialsSecretCAKey] = cluster.CertificateAuthorityData
+	data[corev1alpha1.ResourceCredentialsSecretClientCertKey] = auth.ClientCertificateData
+	data[corev1alpha1.ResourceCredentialsSecretClientKeyKey] = auth.ClientKeyData
+	connectionSecret.Data = data
+
+	return connectionSecret, nil
+}
+
+// wait constants
+const (
+	waitDelay        = 10 * time.Second
+	reasonWaitFailed = "wait failed"
+)
+
+// _wait for cluster running operation (future) to complete
+func (r *Reconciler) _wait(instance *azurecomputev1alpha1.AKSCluster, client azure.AKSClientsetAPI) (reconcile.Result, error) {
+	// check on the status of operation
+	done, err := client.DoneWithContext(ctx, instance.Status.RunningOperation)
+	if err != nil {
+		return r.fail(instance, reasonWaitFailed, err.Error())
+	}
+
+	if !done {
+		r.recorder.Event(instance, corev1.EventTypeNormal, "waiting", "")
+		return reconcile.Result{RequeueAfter: waitDelay}, nil
+	}
+
+	// check the future result to retrieve the detail error if any
+	if rs, err := client.GetResult(instance.Status.RunningOperation); err != nil {
+		// error getting the result (does not mean the result itself is a failure) - requeue
+		r.recorder.Event(instance, corev1.EventTypeWarning, "get result error", err.Error())
+		return resultDone, err
+	} else if rs.StatusCode != http.StatusOK {
+		// result is not successful - dump response body into the event and log
+		b, _ := ioutil.ReadAll(rs.Body)
```

Comment:
were you able to test this out on a real AKS creation failure?  why not directly use the `Result` function instead of handling the reading of the http response body ourselves? https://github.com/Azure/azure-sdk-for-go/blob/master/services/containerservice/mgmt/2018-03-31/containerservice/models.go#L1361

Reply:
- No, I was not able to "fail" cluster on purpose during creation.
- The reason why I am not using `Result` function, because, while it performs similar steps, I don't think it is a good fit, and here is why:
  * We cannot call `Result` function until `future` is "done". Well, we can but then we need to test for errors for "Polling Failure" or "Incomplete Error". For this very reason we are doing this "done" test (almost identical to what is in `Result` function, perhaps, even better since we are not using deprecated `Done()` call) in our `wait` routine. 
  * `Result` returns (or attempts to) a `ManagedCluster` itself, **iff** the `future` response code is `200` or `201`, otherwise it returns an error. Whereas, we need the result of the operation we are waiting for in terms of `success` or `failure`. More specifically we are trying to retrieve the response body **iff** the result was not a "success".

Reply:
In this code here, we are not calling `GetResult` until the future is done, so I don't see why that is an issue.

Won't `Result` give us the response body when the result was not a success?  Isn't that what we want? Why does it matter that it also reads the response body (and tries to marshal to a `ManagedCluster` object) for success?

I'm not understanding why these are reasons to avoid the Azure SDK function altogether and write our own code that we have to take on the maintenance burden for.